### DISCUSSION TITLE
Recommended way to call custom Rails helper methods

### DISCUSSION THREAD

**Discussion by adamlogic** [yippee-fun/phlex#594] [Help] [answered]
When migrating a legacy app to Phlex, we have many existing helper methods that return HTML. I can call these methods via `helpers.my_helper`, but that doesn't append to the output buffer. My current solution is to call `unsafe_raw helpers.my_helper`, and I'm wondering if there's a better way.

```ruby
def template
  div class: "relative px-4" do
    unsafe_raw helpers.svg_icon(:chevronleft)
  end
end
```

**Comment by joeldrapper** [comment, marked as answer]
That’s a really good question. It’s not been considered part of the public API, but there's actually a module to help define adapters for Rails helpers, which is used in defining all the built-in adapters. The built-in adapters are mostly all defined in [this file](https://github.com/phlex-ruby/phlex-rails/blob/main/lib/phlex/rails/helpers.rb), which is a good example to look at.

The module is [`Phlex::Rails::HelperMacros`](https://github.com/phlex-ruby/phlex-rails/blob/main/lib/phlex/rails/helper_macros.rb), and it provides the following:

### `.define_output_helper`

This is for helpers that return HTML. You’d typically use them with `<%= %>` tags in ERB.

**Example:**

```ruby
module Helpers::AudioTag
  extend Phlex::Rails::HelperMacros
  define_output_helper :audio_tag
end
```

### `.define_output_helper_with_capture_block`

This is for output helpers that accept a block, e.g. `link_to("/") { "Hello" }`

**Example:**

```ruby
module Helpers::LinkTo
  extend Phlex::Rails::HelperMacros
  define_output_helper_with_capture_block :link_to
end
```

### `.define_value_helper`

This is for helpers that return a value that you need to use directly. You would typically use these helpers as content or attribute values.

**Example:**

```ruby
module Helpers::AssetURL
  extend Phlex::Rails::HelperMacros
  define_value_helper :asset_url
end
```

### `.define_value_helper_with_capture_block`

Same as value helpers, but where the helper accepts a block.

**Example:**

```ruby
module Helpers::ContentFor
  extend Phlex::Rails::HelperMacros
  define_value_helper_with_capture_block :content_for
end
```

### `.define_builder_yielding_helper`

You probably won’t need this, but these helpers yield a builder object to the given block, and that builder object needs to be decorated so that its methods work as expected.

**Example:**

```ruby
module Helpers::FormWith
  extend Phlex::Rails::HelperMacros
  define_builder_yielding_helper :form_with, Phlex::Rails::BufferedFormBuilder
end
```

Note that `BufferedFormBuilder` is a decorator for a form builder [defined here](https://github.com/phlex-ruby/phlex-rails/blob/main/lib/phlex/rails/buffered_form_builder.rb). It too has methods that can yield builders such as `label`, `fields` and `fields_for`.

![giphy](https://github.com/phlex-ruby/phlex/assets/246692/583b3994-c2d9-456c-a39c-d0cafb864d06)

***

For your use case, you probably want to define a module — maybe `ApplicationHelpers` — that adapts all your custom helpers:

```ruby
module ApplicationHelpers
  extend Phlex::Rails::HelperMacros

  define_output_helper :foo_helper
  define_value_helper :bar_helper
end
```

Then you can include this in `ApplicationComponent`, and the methods `#foo_helper` and `bar_helper` will be available.

```ruby
class ApplicationComponent < Phlex::HTML
  include ApplicationHelpers
end
```

```ruby
class MyComponent < ApplicationComponent
  def template
    foo_helper
    h1(class: bar_helper) { "Hi" }
  end
end
```

**Reply by adamlogic** [reply]
Amazing answer. Thanks!!

**Reply by joeldrapper** [reply]
I should add that it’s perfectly fine to do something like this directly instead, especially if you don’t need to worry about capturing blocks.

```ruby
def svg_icon(...)
  unsafe_raw helpers.svg_icon(...)
end
```

If you want to be extra safe, you could check that the helper returns a safe string before passing it to `unsafe_raw`.

```ruby
def svg_icon(...)
  output = helpers.svg_icon(...)
  
  if ActiveSupport::SafeBuffer === output
    unsafe_raw output
  end
end
```

**Reply by ibrahima** [reply]
Great answer! I wanted to mention that in 2025, the methods are now called [register_*_helper](https://github.com/phlex-ruby/phlex-rails/blob/3f302be12a6ff9dc2fcbcf8cc7513058a18d393c/lib/phlex/rails/helper_macros.rb#L5).

**Comment by mhenrixon** [comment]
Due to the latest updates, I saw that I should no longer use `helpers.` which is fine but the alternative isn't working as I want it to.

```ruby
class Views::AdminBase < Views::Base
  include Phlex::Rails::Helpers::SearchField

  register_value_helper :search_form_for
  register_output_helper :sort_link
end
```

The search form for is a ransack thing. When trying to use it, it acts weird:

```
 def view_template
    page_title "Users"
    headline "Users"

    div(class: "content") do
      search_form_for @query, url: admin_users_path do |f|
        table(class: "table") do
          tbody do
            @users.find_each do |user|
              render Views::Admin::Users::User.new(user:)
            end
          end
        end

        f.submit "Filter", class: "hidden" # this raises an error:
      end
      render_pagy_nav
    end
  end
```

![CleanShot 2025-02-16 at 16 41 06@2x](https://github.com/user-attachments/assets/777af091-e086-48c0-ae68-a301e0fecc62)

If I use helpers.search_form_for it works as expected. Any suggestions?

**Reply by joeldrapper** [reply]
Unfortunately there is no easy way to adapt helpers that yield form builders. We’ve done it for `form_with` and `form_for`, but it’s not something you can automate.

We should probably open an issue in the Ransack repo to discuss this further. A temporary solution that will get you unblocked is just moving the search form to a partial. You can then render that partial in Phlex.

Something like this:

```ruby
render partial("search_form")
```

And in your `_search_form.html.erb` partial:

```erb
<%= search_form_for @query, url: admin_users_path do |f| %>
  <table class="table">
    <tbody> do
      <% @users.find_each do |user| %>
        <%= render Views::Admin::Users::User.new(user:) %>
      <% end %>
    </tbody>
  </table>

  <%= f.submit "Filter", class: "hidden" %>
<% end %>
```

**Reply by mhenrixon** [reply]
That's a workaround! I'll do that for now but all these things add up so great to have in the future. 

I'll check the form_with helper as well.
